### PR TITLE
[HigherOrderOp] Support SymInt as input to body function

### LINKED ISSUE #108283: [HigherOrderOp] Should support SymInt as input to body function
### 🐛 Describe the bug

Right now, if a SymInt gets passed to the body of a higher-order op, we don't support it:

```py
import torch
from torch._higher_order_ops.wrap import wrap

@torch.compile(backend='eager', fullgraph=True, dynamic=True)
def f(x):
    i = x.size(0)
    return wrap(lambda x, i: x.view(i), x, i)

x = torch.randn(3, 1)
f(x)
```
(try dynamic=False to see it succeed). We should support this; the symint should get recorded as the input of the body function.

NB: wrap(f, *args) is a higher order op whose semantics are to just run f(*args).

### Error logs

_No response_

### Minified repro

_No response_

### Versions

main

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #108283

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -218,6 +218,16 @@ def f(x):
...
 
+    def test_symint_input(self):
+        def f(x):
+            i = x.size(0)
+            return wrap(lambda x, i: x.view(i), x, i)
+
+        x = torch.randn(3, 1)
+        self._test_wrap_simple(
+            f, (x,), ifdynstaticdefault(2, 3), expected_opcount=ifdynstaticdefault(1, 2)
+        )
```
Comment (yanboliang) [inline test/dynamo/test_higher_order_ops.py:229]: The corresponding dynamic shape version test was generated at ```test/dynamo/test_dynamic_shapes.py``` automatically, so we don't need to explicitly set ```dynamic_shape=True``` at here. And both static and dynamic shapes will be tested.

File: test/dynamo/test_higher_order_ops.py [hunk 2]
```diff
@@ -969,6 +979,27 @@ def f(x):
...
+        y = torch.randn(3, 1)
+        compiled_fn = torch.compile(fn, backend=cnt, fullgraph=False)
+
+        ref = fn(x, y)
+        res = compiled_fn(x, y)
+
+        self.assertEqual(ref, res)
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(cnt.op_count, ifdynstaticdefault(2, 3))
+
```
Comment (yanboliang) [inline test/dynamo/test_higher_order_ops.py:1002]: I don't know which higher order ops should support SymInt as input, ```cond``` has restrictions that all inputs should be tensor, so I randomly pick ```map``` to test.

File: torch/_dynamo/variables/higher_order_ops.py [hunk 3]
```diff
@@ -95,6 +100,17 @@ def validate_args_and_maybe_create_graph_inputs(
...
+            if manually_set_subgraph_inputs:
+                new_proxy = tracer.create_graph_input(str(a.sym_num.node.expr))
+                new_arg = wrap_fx_proxy_cls(
+                    target_cls=SymNodeVariable,
+                    tx=tx,
+                    proxy=new_proxy,
+                    example_value=a.sym_num,
+                )
+            else:
+                new_arg = a
```
Comment (yanboliang) [inline torch/_dynamo/variables/higher_order_ops.py:113]: Both branches are tested:
* ```manually_set_subgraph_inputs=True``` -> ```map```
* ```manually_set_subgraph_inputs=False``` -> ```wrap```

File: test/dynamo/test_higher_order_ops.py (hunk 2 above)
Comment (zou3519) [reply test/dynamo/test_higher_order_ops.py:966]: That's fine

File: torch/_dynamo/variables/higher_order_ops.py (hunk 3 above)
Comment (zou3519) [reply torch/_dynamo/variables/higher_order_ops.py:113]: Thanks for pointing this out

File: test/dynamo/test_higher_order_ops.py [hunk 4]
```diff
@@ -969,6 +943,27 @@ def f(x):
...
+        def fn(x, y):
+            def inner(x, y):
+                return torch.sin(x + y)
+
+            return control_flow.map(inner, x, y.size(0))
+
+        x = torch.randn(3, 1)
+        y = torch.randn(3, 1)
+        compiled_fn = torch.compile(fn, backend=cnt, fullgraph=False)
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:958]: Why fullgraph=False? I would expect with these changes we can fullgraph=True

**Review by zou3519 — APPROVED** [review]
LGTM, minor comment

File: test/dynamo/test_higher_order_ops.py (hunk 4 above)
Comment (yanboliang) [reply test/dynamo/test_higher_order_ops.py:958]: It's a typo, will update it.

**Pull request closed without merging**